pkg/controller/infrastructure/dualstack/dualstack.go

```diff
@@ -0,0 +1,112 @@
+package dualstack
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/gardener/gardener-extension-provider-alicloud/pkg/alicloud"
+	alicloudclient "github.com/gardener/gardener-extension-provider-alicloud/pkg/alicloud/client"
+)
+
+// DualStack contains values of EIP used to render terraform charts
+type DualStack struct {
+	Enabled            bool
+	Zone_A             string
+	Zone_A_CIDR        string
+	Zone_A_IPV6_SUBNET int
+	Zone_B             string
+	Zone_B_CIDR        string
+	Zone_B_IPV6_SUBNET int
+}
+
+// CreateDualStackValues create DualStack values
+func CreateDualStackValues(
+	enableDualStack bool,
+	region string,
+	vpcCidr *string,
+	credentials *alicloud.Credentials,
+	clientFactory alicloudclient.ClientFactory,
+) (*DualStack, error) {
+	dualStack := DualStack{
+		Enabled: enableDualStack,
+	}
+	if !enableDualStack {
+		return &dualStack, nil
+	}
+	if vpcCidr == nil {
+		return nil, fmt.Errorf("vpcCidr must be set")
+	}
+	dualStack.Zone_A_IPV6_SUBNET = 255
+	dualStack.Zone_B_IPV6_SUBNET = 254
+	nlbClient, err := clientFactory.NewNLBClient(region, credentials.AccessKeyID, credentials.AccessKeySecret)
+	if err != nil {
+		return nil, err
+	}
+	zones, err := nlbClient.GetNLBAvailableZones(region)
+	if err != nil {
+		return nil, err
+	}
+	if len(zones) < 2 {
+		return nil, fmt.Errorf("not enough available zones for DualStack")
+	}
+	dualStack.Zone_A = zones[0]
+	dualStack.Zone_B = zones[1]
+
+	// DualStack only for managed vpc
+
+	subCidrs, err := getLastSubCidr(*vpcCidr, 28, 2)
+	if err != nil || subCidrs == nil || len(subCidrs) < 2 {
+		return nil, fmt.Errorf("get sub cidr failed")
+	}
+
+	dualStack.Zone_A_CIDR = subCidrs[0]
+	dualStack.Zone_B_CIDR = subCidrs[1]
+
+	return &dualStack, nil
+}
+
+func getLastSubCidr(originCidr string, subNetMaskLen, count int) ([]string, error) {
+	_, ipnet, err := net.ParseCIDR(originCidr)
+	if err != nil {
+		return nil, err
+	}
+	if count <= 0 {
+		count = 1
+	}
+
+	orgin_net_mask_len, _ := ipnet.Mask.Size()
+	if orgin_net_mask_len > subNetMaskLen {
+		return nil, fmt.Errorf("not enough capacity to divide sub cidr")
+	}
+	sub_ip_mask_len := 32 - subNetMaskLen
+
+	subnets := 1 << uint(32-orgin_net_mask_len-sub_ip_mask_len)
```

Comment:
This seems to be pure IPv4 logic, but we have not even checked whether the `originCidr` is IPv4. Is that always ensured somehow or should we simply add a check?
WDYT?

Reply:
This is a local function and the input param is IPV4 address. I've changed the function name to getLastIpv4SubCidr